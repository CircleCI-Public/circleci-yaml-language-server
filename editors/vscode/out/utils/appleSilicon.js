"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAppleSilicon = void 0;
const os = require("os");
/**
 * Check if the user is running on an Apple Silicon chipset
 * @returns true if the processor is an Apple one (such as the M1)
 */
function isAppleSilicon() {
    return os.cpus().some((cpu) => cpu.model.startsWith('Apple '));
}
exports.isAppleSilicon = isAppleSilicon;
//# sourceMappingURL=appleSilicon.js.map