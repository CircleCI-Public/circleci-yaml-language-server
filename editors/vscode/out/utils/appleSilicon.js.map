{"version":3,"file":"appleSilicon.js","sourceRoot":"","sources":["../../src/utils/appleSilicon.ts"],"names":[],"mappings":";;;AAAA,yBAAyB;AAEzB;;;GAGG;AACH,SAAgB,cAAc;IAC1B,OAAO,EAAE,CAAC,IAAI,EAAE,CAAC,IAAI,CAAC,CAAC,GAAG,EAAE,EAAE,CAAC,GAAG,CAAC,KAAK,CAAC,UAAU,CAAC,QAAQ,CAAC,CAAC,CAAC;AACnE,CAAC;AAFD,wCAEC"}