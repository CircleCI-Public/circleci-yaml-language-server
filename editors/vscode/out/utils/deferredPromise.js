"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDeferredPromise = void 0;
function createDeferredPromise() {
    let resolve;
    let reject;
    const promise = new Promise((res, rej) => {
        resolve = res;
        reject = rej;
    });
    let created = Date.now();
    //@ts-ignore
    return { resolve, reject, promise, created };
}
exports.createDeferredPromise = createDeferredPromise;
//# sourceMappingURL=deferredPromise.js.map