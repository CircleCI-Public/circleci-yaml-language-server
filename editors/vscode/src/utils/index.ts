export * from './deferredPromise';
export * from './outputChannel';
export * from './appleSilicon';
export * from './devMode';
