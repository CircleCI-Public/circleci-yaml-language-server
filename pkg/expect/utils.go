package expect

import (
	"fmt"

	"github.com/CircleCI-Public/circleci-yaml-language-server/pkg/utils"
	"go.lsp.dev/protocol"
)

// Indicate if a diagnostic is present in the given list
// Will return true if a diagnostic match all the given conditions:
//   - Same code
//   - Same message
//   - Same range
//   - Same severity
func ListHasDiagnostic(list []protocol.Diagnostic, diagnostic protocol.Diagnostic) bool {
	if len(list) == 0 {
		return false
	}

	for _, diag := range list {
		if AreDiagnosticEquivalent(diag, diagnostic) {
			return true
		}
	}

	return false
}

// Compare two diagnostics and indicate if they are equivalents or not
// Two diagnostics are equivalent if they match all of the following conditions:
//   - Same code
//   - Same message
//   - Same range
//   - Same severity
func AreDiagnosticEquivalent(a protocol.Diagnostic, b protocol.Diagnostic) bool {
	if a.Severity != b.Severity {
		return false
	}

	if a.Message != b.Message {
		return false
	}

	if a.Code != b.Code {
		return false
	}

	return utils.AreRangeEqual(a.Range, b.Range)
}

// Return a string displaying a list of diagnostic information.
func diagnosticInfoList(list []protocol.Diagnostic, prefix string) string {
	listStr := ""
	for _, diag := range list {
		listStr += prefix + diagnosticInfo(diag)
	}

	return listStr
}

// Return a string containing the diagnostic information.
func diagnosticInfo(diagnostic protocol.Diagnostic) string {
	return fmt.Sprintf(
		"<L%d:%d,L%d:%d> %s: %s",
		diagnostic.Range.Start.Line,
		diagnostic.Range.Start.Character,
		diagnostic.Range.End.Line,
		diagnostic.Range.End.Character,
		diagnostic.Severity,
		diagnostic.Message,
	)
}
