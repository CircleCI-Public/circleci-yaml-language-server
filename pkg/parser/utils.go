package parser

import (
	"fmt"
	"strconv"
	"strings"

	sitter "github.com/smacker/go-tree-sitter"
	ymlgrammar "github.com/smacker/go-tree-sitter/yaml"
	"go.lsp.dev/protocol"
)

func GetRootNode(content []byte) *sitter.Node {
	parser := sitter.NewParser()
	parser.SetLanguage(ymlgrammar.GetLanguage())

	tree := parser.Parse(nil, content)

	return tree.RootNode()
}

func GetChildOfType(node *sitter.Node, typeName string) *sitter.Node {
	if node == nil {
		return nil
	}
	for i := 0; uint32(i) < node.ChildCount(); i++ {
		child := node.Child(i)
		if child.Type() == typeName {
			return child
		}
	}
	return nil
}

func GetChildMapping(node *sitter.Node) *sitter.Node {
	blockMappingNode := GetChildOfType(node, "block_mapping")

	if blockMappingNode != nil {
		return blockMappingNode
	}

	return GetChildOfType(node, "flow_mapping")
}

func GetBlockMappingNode(streamNode *sitter.Node) *sitter.Node {
	documentNode := GetChildOfType(streamNode, "document")
	if documentNode.Type() != "document" {
		return nil
	}
	blockNode := GetChildOfType(documentNode, "block_node")
	if blockNode == nil {
		return nil
	}

	return GetChildOfType(blockNode, "block_mapping")
}

func (doc *YamlDocument) GetNodeText(node *sitter.Node) string {
	res := doc.GetRawNodeText(node)

	if strings.HasPrefix(res, "\"") && strings.HasSuffix(res, "\"") {
		res = strings.Trim(res, "\"")
	} else if strings.HasPrefix(res, "'") && strings.HasSuffix(res, "'") {
		res = strings.Trim(res, "'")
	}

	res = strings.TrimPrefix(res, "|\n")
	res = strings.TrimSpace(res)

	return res
}

func (doc *YamlDocument) GetRawNodeText(node *sitter.Node) string {
	if node == nil {
		return ""
	}
	res := string(doc.Content[node.StartByte():node.EndByte()])
	return res
}

type TextAndRange struct {
	Text  string
	Range protocol.Range
}

func (doc *YamlDocument) getNodeTextArray(valueNode *sitter.Node) []string {
	res := doc.getNodeTextArrayWithRange(valueNode)
	texts := make([]string, len(res))
	for i, textAndRange := range res {
		texts[i] = textAndRange.Text
	}
	return texts
}

func (doc *YamlDocument) getNodeTextArrayWithRange(valueNode *sitter.Node) []TextAndRange {
	// valueNode is block_node which has a block_sequence child
	blockSequenceNode := GetChildOfType(valueNode, "block_sequence")
	texts := make([]TextAndRange, 0)

	if blockSequenceNode == nil {
		blockSequenceNode = GetChildOfType(valueNode, "flow_sequence")
		if blockSequenceNode == nil {
			return texts
		}
	}

	iterateOnBlockSequence(blockSequenceNode, func(child *sitter.Node) {
		// If blockSequence is a flow_sequence, then the child is
		// directly a flow_node
		if child.Type() == "flow_node" {
			texts = append(texts, TextAndRange{Text: doc.GetNodeText(child), Range: NodeToRange(child)})
		} else {
			// But if the blockSequence is a block_sequence, then the child is
			// a block_sequence_item
			element := GetChildOfType(child, "flow_node")
			hyphenNode := child.Child(0)
			if element != nil {
				texts = append(texts, TextAndRange{Text: doc.GetNodeText(element), Range: NodeToRange(child)})
			} else if hyphenNode != nil {
				texts = append(texts, TextAndRange{
					Text: doc.GetNodeText(element),
					Range: protocol.Range{
						Start: protocol.Position{
							Line:      hyphenNode.StartPoint().Row,
							Character: hyphenNode.StartPoint().Column + 1,
						},
						End: protocol.Position{
							Line:      hyphenNode.EndPoint().Row,
							Character: hyphenNode.EndPoint().Column + 2,
						},
					},
				})
			}
		}
	})

	return texts
}

func (doc *YamlDocument) getNodeTextArrayOrText(valueNode *sitter.Node) []string {
	textArray := doc.getNodeTextArray(valueNode)
	if len(textArray) == 0 {
		return []string{doc.GetNodeText(valueNode)}
	}
	return textArray
}

func (doc *YamlDocument) parseDictionary(valueNode *sitter.Node) map[string]string {
	dictionary := make(map[string]string)

	iterateOnBlockMapping(valueNode, func(child *sitter.Node) {
		if child.Type() == "block_mapping_pair" || child.Type() == "flow_pair" {
			key := child.ChildByFieldName("key")
			value := child.ChildByFieldName("value")
			if key != nil && value != nil {
				dictionary[doc.GetNodeText(key)] = doc.GetNodeText(value)
			}
		}
	})

	return dictionary
}

func (doc *YamlDocument) parseDescription(descriptionNode *sitter.Node) string {
	return doc.GetNodeText(descriptionNode)
}

func iterateOnBlockMapping(blockMappingNode *sitter.Node, fn func(child *sitter.Node)) {
	if blockMappingNode == nil || (blockMappingNode.Type() != "block_mapping" && blockMappingNode.Type() != "flow_mapping") {
		return
	}

	for i := 0; uint32(i) < blockMappingNode.ChildCount(); i++ {
		child := blockMappingNode.Child(i)

		if child.Type() == "comment" {
			continue
		}

		fn(child)
	}
}

func iterateOnBlockSequence(blockSequenceNode *sitter.Node, fn func(child *sitter.Node)) {
	if blockSequenceNode == nil ||
		(blockSequenceNode.Type() != "block_sequence" && blockSequenceNode.Type() != "flow_sequence") {
		return
	}
	for i := 0; uint32(i) < blockSequenceNode.ChildCount(); i++ {
		child := blockSequenceNode.Child(i)

		if child.Type() == "comment" {
			continue
		}

		fn(child)
	}
}

func ExecQuery(node *sitter.Node, query string, fn func(match *sitter.QueryMatch)) error {
	pattern := []byte(query)
	queryTreeSitter, err := sitter.NewQuery(pattern, ymlgrammar.GetLanguage())
	if err != nil {
		return err
	}

	cursor := sitter.NewQueryCursor()
	cursor.Exec(queryTreeSitter, node)
	anchorMatches, found := cursor.NextMatch()
	for found {
		fn(anchorMatches)
		anchorMatches, found = cursor.NextMatch()
	}

	return nil
}

func FindDeepestNode(rootNode *sitter.Node, content []byte, toFind []string) (*sitter.Node, error) {
	if len(toFind) == 0 {
		return rootNode, nil
	}

	iterator := sitter.NewIterator(rootNode, sitter.DFSMode)
	node, err := iterator.Next()

	for err == nil {
		if intValue, err := strconv.Atoi(toFind[0]); err == nil {
			if node.Type() == "block_sequence" {
				if node.ChildCount() < uint32(intValue+1) {
					return nil, fmt.Errorf("index out of range: trying to access %d in array of size %d", intValue, node.ChildCount())
				}

				childNode := node.Child((intValue))
				return FindDeepestNode(childNode, content, toFind[1:])
			}
		}
		if node.Type() == "block_mapping_pair" {
			if key := node.ChildByFieldName("key"); string(content[key.StartByte():key.EndByte()]) == toFind[0] {
				return FindDeepestNode(node, content, toFind[1:])
			}
		}
		node, err = iterator.Next()
	}

	return node, fmt.Errorf("not found")
}

func NodeToRange(node *sitter.Node) protocol.Range {
	return protocol.Range{
		Start: protocol.Position{Line: node.StartPoint().Row, Character: node.StartPoint().Column},
		End:   protocol.Position{Line: node.EndPoint().Row, Character: node.EndPoint().Column},
	}
}

func getKeyValueNodes(node *sitter.Node) (keyNode *sitter.Node, valueNode *sitter.Node) {
	if node != nil && (node.Type() == "block_mapping_pair" || node.Type() == "flow_pair") {
		keyNode = node.ChildByFieldName("key")
		valueNode = node.ChildByFieldName("value")
	}
	return
}
