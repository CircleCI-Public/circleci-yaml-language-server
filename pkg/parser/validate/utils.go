package validate

import (
	"fmt"

	"github.com/circleci/circleci-yaml-language-server/pkg/ast"
	"github.com/circleci/circleci-yaml-language-server/pkg/utils"
	"go.lsp.dev/protocol"
)

func (val Validate) createParameterError(param ast.ParameterValue, stepName string, shouldBeType string) {
	*val.Diagnostics = append(*val.Diagnostics, utils.CreateErrorDiagnosticFromRange(
		param.Range,
		fmt.Sprintf("Parameter %s for %s must be a %s", param.Name, stepName, shouldBeType)),
	)
}

func (val Validate) addDiagnostic(diagnostic protocol.Diagnostic) {
	*val.Diagnostics = append(*val.Diagnostics, diagnostic)
}
