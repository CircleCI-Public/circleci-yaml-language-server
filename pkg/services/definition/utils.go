package definition

import (
	"fmt"

	yamlparser "github.com/CircleCI-Public/circleci-yaml-language-server/pkg/parser"
	sitter "github.com/smacker/go-tree-sitter"
	"go.lsp.dev/protocol"
)

func (def DefinitionStruct) getCommandOrJobLocation(name string, includeCommands bool) ([]protocol.Location, error) {
	if job, ok := def.Doc.Jobs[name]; ok {
		return []protocol.Location{
			{
				Range: job.Range,
				URI:   def.Doc.URI,
			},
		}, nil
	}

	if command, ok := def.Doc.Commands[name]; ok && includeCommands {
		return []protocol.Location{
			{
				Range: command.Range,
				URI:   def.Doc.URI,
			},
		}, nil
	}

	if orb, err := def.getOrbLocation(name, true); err == nil {
		return orb, nil
	}

	return []protocol.Location{}, fmt.Errorf("command or job not found")
}

func (def DefinitionStruct) getCommandOrJobParamLocation(name string, paramName string, includeCommands bool) ([]protocol.Location, error) {
	if job, ok := def.Doc.Jobs[name]; ok {
		if param, ok := job.Parameters[paramName]; ok {
			return []protocol.Location{
				{
					Range: param.GetRange(),
					URI:   def.Doc.URI,
				},
			}, nil
		}
	}

	if command, ok := def.Doc.Commands[name]; ok && includeCommands {
		if param, ok := command.Parameters[paramName]; ok {
			return []protocol.Location{
				{
					Range: param.GetRange(),
					URI:   def.Doc.URI,
				},
			}, nil
		}
	}

	if orb, err := def.getOrbParamLocation(name, paramName); err == nil {
		return orb, nil
	}

	return []protocol.Location{}, fmt.Errorf("command or job not found")
}

func GetPathFromVisitedNodes(visitedNodes []*sitter.Node, doc yamlparser.YamlDocument) []string {
	var path []string
	if len(visitedNodes) == 0 {
		return path
	}

	for _, node := range visitedNodes[1:] {
		switch node.Type() {
		case "block_mapping_pair":
			key := node.ChildByFieldName("key")
			name := string(doc.Content[key.StartByte():key.EndByte()])
			path = append(path, name)
		case "block_sequence_item":
			flow_node := yamlparser.GetChildOfType(node, "flow_node")
			if flow_node != nil {
				name := string(doc.Content[flow_node.StartByte():flow_node.EndByte()])
				path = append(path, name)
			}
		}
	}

	return path
}
