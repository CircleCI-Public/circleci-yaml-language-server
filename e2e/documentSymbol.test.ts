import { commands } from './utils';

describe('DocumentSymbol', () => {
  it('Standard file', async () => {
    const testingFile = 'config1.yml';
    await commands.didOpen(testingFile);

    const res = await commands.documentSymbol(testingFile);

    expect(res).toMatchSnapshot();
  });

  it('File with continuation workflow', async () => {
    const testingFile = 'continuation.yml';
    await commands.didOpen(testingFile);

    const res = await commands.documentSymbol(testingFile);

    expect(res).toMatchSnapshot();
  });
});
