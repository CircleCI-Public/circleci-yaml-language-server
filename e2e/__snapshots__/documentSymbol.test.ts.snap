// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentSymbol File with continuation workflow 1`] = `
[
  {
    "detail": "2.1",
    "kind": 0,
    "name": "Version",
    "range": {
      "end": {
        "character": 12,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "selectionRange": {
      "end": {
        "character": 12,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
  },
  {
    "detail": "true",
    "kind": 17,
    "name": "Setup",
    "range": {
      "end": {
        "character": 11,
        "line": 2,
      },
      "start": {
        "character": 0,
        "line": 2,
      },
    },
    "selectionRange": {
      "end": {
        "character": 11,
        "line": 2,
      },
      "start": {
        "character": 0,
        "line": 2,
      },
    },
  },
  {
    "children": [
      {
        "detail": "0.1.2",
        "kind": 12,
        "name": "continuation",
        "range": {
          "end": {
            "character": 43,
            "line": 5,
          },
          "start": {
            "character": 2,
            "line": 5,
          },
        },
        "selectionRange": {
          "end": {
            "character": 43,
            "line": 5,
          },
          "start": {
            "character": 2,
            "line": 5,
          },
        },
      },
    ],
    "kind": 18,
    "name": "Orbs",
    "range": {
      "end": {
        "character": 43,
        "line": 5,
      },
      "start": {
        "character": 2,
        "line": 5,
      },
    },
    "selectionRange": {
      "end": {
        "character": 43,
        "line": 5,
      },
      "start": {
        "character": 2,
        "line": 5,
      },
    },
  },
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "kind": 1,
                "name": "checkout",
                "range": {
                  "end": {
                    "character": 16,
                    "line": 11,
                  },
                  "start": {
                    "character": 8,
                    "line": 11,
                  },
                },
                "selectionRange": {
                  "end": {
                    "character": 16,
                    "line": 11,
                  },
                  "start": {
                    "character": 8,
                    "line": 11,
                  },
                },
              },
              {
                "kind": 1,
                "name": "Generate config",
                "range": {
                  "end": {
                    "character": 11,
                    "line": 12,
                  },
                  "start": {
                    "character": 8,
                    "line": 12,
                  },
                },
                "selectionRange": {
                  "end": {
                    "character": 11,
                    "line": 12,
                  },
                  "start": {
                    "character": 8,
                    "line": 12,
                  },
                },
              },
              {
                "kind": 1,
                "name": "continuation/continue",
                "range": {
                  "end": {
                    "character": 29,
                    "line": 16,
                  },
                  "start": {
                    "character": 8,
                    "line": 16,
                  },
                },
                "selectionRange": {
                  "end": {
                    "character": 29,
                    "line": 16,
                  },
                  "start": {
                    "character": 8,
                    "line": 16,
                  },
                },
              },
            ],
            "detail": "3 total",
            "kind": 18,
            "name": "Steps",
            "range": {
              "end": {
                "character": 50,
                "line": 17,
              },
              "start": {
                "character": 4,
                "line": 10,
              },
            },
            "selectionRange": {
              "end": {
                "character": 50,
                "line": 17,
              },
              "start": {
                "character": 4,
                "line": 10,
              },
            },
          },
          {
            "kind": 6,
            "name": "Executor: continuation/default",
            "range": {
              "end": {
                "character": 34,
                "line": 9,
              },
              "start": {
                "character": 4,
                "line": 9,
              },
            },
            "selectionRange": {
              "end": {
                "character": 34,
                "line": 9,
              },
              "start": {
                "character": 4,
                "line": 9,
              },
            },
          },
        ],
        "kind": 1,
        "name": "setup",
        "range": {
          "end": {
            "character": 50,
            "line": 17,
          },
          "start": {
            "character": 2,
            "line": 8,
          },
        },
        "selectionRange": {
          "end": {
            "character": 50,
            "line": 17,
          },
          "start": {
            "character": 2,
            "line": 8,
          },
        },
      },
    ],
    "kind": 18,
    "name": "Jobs",
    "range": {
      "end": {
        "character": 50,
        "line": 17,
      },
      "start": {
        "character": 2,
        "line": 8,
      },
    },
    "selectionRange": {
      "end": {
        "character": 50,
        "line": 17,
      },
      "start": {
        "character": 2,
        "line": 8,
      },
    },
  },
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "kind": 0,
                "name": "setup",
                "range": {
                  "end": {
                    "character": 13,
                    "line": 22,
                  },
                  "start": {
                    "character": 6,
                    "line": 22,
                  },
                },
                "selectionRange": {
                  "end": {
                    "character": 13,
                    "line": 22,
                  },
                  "start": {
                    "character": 6,
                    "line": 22,
                  },
                },
              },
            ],
            "kind": 18,
            "name": "Jobs",
            "range": {
              "end": {
                "character": 0,
                "line": 23,
              },
              "start": {
                "character": 4,
                "line": 21,
              },
            },
            "selectionRange": {
              "end": {
                "character": 0,
                "line": 23,
              },
              "start": {
                "character": 4,
                "line": 21,
              },
            },
          },
        ],
        "kind": 23,
        "name": "setup",
        "range": {
          "end": {
            "character": 0,
            "line": 23,
          },
          "start": {
            "character": 2,
            "line": 20,
          },
        },
        "selectionRange": {
          "end": {
            "character": 0,
            "line": 23,
          },
          "start": {
            "character": 2,
            "line": 20,
          },
        },
      },
    ],
    "kind": 18,
    "name": "Workflows",
    "range": {
      "end": {
        "character": 0,
        "line": 23,
      },
      "start": {
        "character": 2,
        "line": 20,
      },
    },
    "selectionRange": {
      "end": {
        "character": 0,
        "line": 23,
      },
      "start": {
        "character": 2,
        "line": 20,
      },
    },
  },
]
`;

exports[`DocumentSymbol Standard file 1`] = `
[
  {
    "detail": "2.1",
    "kind": 0,
    "name": "Version",
    "range": {
      "end": {
        "character": 12,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "selectionRange": {
      "end": {
        "character": 12,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
  },
  {
    "children": [
      {
        "detail": "2",
        "kind": 12,
        "name": "macos",
        "range": {
          "end": {
            "character": 25,
            "line": 3,
          },
          "start": {
            "character": 2,
            "line": 3,
          },
        },
        "selectionRange": {
          "end": {
            "character": 25,
            "line": 3,
          },
          "start": {
            "character": 2,
            "line": 3,
          },
        },
      },
    ],
    "kind": 18,
    "name": "Orbs",
    "range": {
      "end": {
        "character": 25,
        "line": 3,
      },
      "start": {
        "character": 2,
        "line": 3,
      },
    },
    "selectionRange": {
      "end": {
        "character": 25,
        "line": 3,
      },
      "start": {
        "character": 2,
        "line": 3,
      },
    },
  },
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "kind": 1,
                "name": "checkout",
                "range": {
                  "end": {
                    "character": 18,
                    "line": 13,
                  },
                  "start": {
                    "character": 10,
                    "line": 13,
                  },
                },
                "selectionRange": {
                  "end": {
                    "character": 18,
                    "line": 13,
                  },
                  "start": {
                    "character": 10,
                    "line": 13,
                  },
                },
              },
              {
                "kind": 1,
                "name": "run",
                "range": {
                  "end": {
                    "character": 13,
                    "line": 14,
                  },
                  "start": {
                    "character": 10,
                    "line": 14,
                  },
                },
                "selectionRange": {
                  "end": {
                    "character": 13,
                    "line": 14,
                  },
                  "start": {
                    "character": 10,
                    "line": 14,
                  },
                },
              },
              {
                "kind": 1,
                "name": "macos/switch-ruby",
                "range": {
                  "end": {
                    "character": 27,
                    "line": 15,
                  },
                  "start": {
                    "character": 10,
                    "line": 15,
                  },
                },
                "selectionRange": {
                  "end": {
                    "character": 27,
                    "line": 15,
                  },
                  "start": {
                    "character": 10,
                    "line": 15,
                  },
                },
              },
            ],
            "detail": "3 total",
            "kind": 18,
            "name": "Steps",
            "range": {
              "end": {
                "character": 24,
                "line": 16,
              },
              "start": {
                "character": 4,
                "line": 12,
              },
            },
            "selectionRange": {
              "end": {
                "character": 24,
                "line": 16,
              },
              "start": {
                "character": 4,
                "line": 12,
              },
            },
          },
          {
            "children": [
              {
                "kind": 8,
                "name": "cimg/node:17.2.0",
                "range": {
                  "end": {
                    "character": 31,
                    "line": 8,
                  },
                  "start": {
                    "character": 8,
                    "line": 8,
                  },
                },
                "selectionRange": {
                  "end": {
                    "character": 31,
                    "line": 8,
                  },
                  "start": {
                    "character": 8,
                    "line": 8,
                  },
                },
              },
            ],
            "kind": 13,
            "name": "Docker",
            "range": {
              "end": {
                "character": 24,
                "line": 16,
              },
              "start": {
                "character": 2,
                "line": 6,
              },
            },
            "selectionRange": {
              "end": {
                "character": 24,
                "line": 16,
              },
              "start": {
                "character": 2,
                "line": 6,
              },
            },
          },
        ],
        "kind": 1,
        "name": "build",
        "range": {
          "end": {
            "character": 24,
            "line": 16,
          },
          "start": {
            "character": 2,
            "line": 6,
          },
        },
        "selectionRange": {
          "end": {
            "character": 24,
            "line": 16,
          },
          "start": {
            "character": 2,
            "line": 6,
          },
        },
      },
    ],
    "kind": 18,
    "name": "Jobs",
    "range": {
      "end": {
        "character": 24,
        "line": 16,
      },
      "start": {
        "character": 2,
        "line": 6,
      },
    },
    "selectionRange": {
      "end": {
        "character": 24,
        "line": 16,
      },
      "start": {
        "character": 2,
        "line": 6,
      },
    },
  },
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "kind": 0,
                "name": "build",
                "range": {
                  "end": {
                    "character": 13,
                    "line": 21,
                  },
                  "start": {
                    "character": 6,
                    "line": 21,
                  },
                },
                "selectionRange": {
                  "end": {
                    "character": 13,
                    "line": 21,
                  },
                  "start": {
                    "character": 6,
                    "line": 21,
                  },
                },
              },
            ],
            "kind": 18,
            "name": "Jobs",
            "range": {
              "end": {
                "character": 0,
                "line": 24,
              },
              "start": {
                "character": 4,
                "line": 20,
              },
            },
            "selectionRange": {
              "end": {
                "character": 0,
                "line": 24,
              },
              "start": {
                "character": 4,
                "line": 20,
              },
            },
          },
        ],
        "kind": 23,
        "name": "test-build",
        "range": {
          "end": {
            "character": 0,
            "line": 24,
          },
          "start": {
            "character": 2,
            "line": 19,
          },
        },
        "selectionRange": {
          "end": {
            "character": 0,
            "line": 24,
          },
          "start": {
            "character": 2,
            "line": 19,
          },
        },
      },
    ],
    "kind": 18,
    "name": "Workflows",
    "range": {
      "end": {
        "character": 0,
        "line": 24,
      },
      "start": {
        "character": 2,
        "line": 19,
      },
    },
    "selectionRange": {
      "end": {
        "character": 0,
        "line": 24,
      },
      "start": {
        "character": 2,
        "line": 19,
      },
    },
  },
]
`;
